Improve performance of inbound lanes by making spinning loops in AeronSource configurable

jrudolph opened the pull request:
This adds a new setting `advanced.inbound-source-spinning` which can be used to configure the number of spin loops in AeronSource. Using that setting (and slightly improved benchmarking infrastructure), I ran a set of benchmarks with the results attached.

The gist of it is that removing the spinning completely has mostly positive results with only a few small disadvantages. 

The biggest benefit of disabling spinning is that with `inbound-lanes > 1` latency is now only slightly worse than with 1 lane (66µs with 1 lane vs 76µs with 2 lanes in the 20000k case) and throughput improves by 25% (236k msgs/s with 1 lane vs. 300k msgs/s with 2 lanes) in the 5-to-5 case.

The biggest adverse effect I found was that when `inbound-lanes = 1` (no async boundary), latency in the 20000k/s case (which is the one with the overall best latency) takes a 25% hit from 55µs to 65µs when spinning is disabled (at the same time throughput is better). This might seem much relatively but in absolute numbers it is negligible imo because latency usually is more in the ranges of 100-200µs. 

[benchmark-results.zip](https://github.com/akka/akka/files/664265/benchmark-results.zip)

I added the extra setting mostly for allowing the benchmarks. I'd prefer to remove it in the end. We could also keep it for a while if third-parties would be interested in running their own tests.

jrudolph committed:
=rem #21365 use default directory for shared media driver to /dev/shm

It was reported that shared media driver performance can depend on the
kind of file-system where the files are contained. /dev/shm is an in-memory
filesystem that was reported to work well with the shared aeron media driver.

jrudolph committed:
=rem #21365 cleanup multiple lane creation slightly

ktoso commented on akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/FileReporter.scala:
```diff
@@ -0,0 +1,70 @@
+
+      def close(): Unit = fos.close()
+    }
+}
```
nice one

ktoso commented on akka-remote/src/main/scala/akka/remote/artery/ArterySettings.scala:
```diff
@@ -115,6 +115,9 @@ private[akka] final class ArterySettings private (config: Config) {
       interval > Duration.Zero, "inbound-restart-timeout must be more than zero")
     val InboundMaxRestarts = getInt("inbound-max-restarts")
+    val InboundSourceSpinning = getInt("inbound-source-spinning").requiring(spins ⇒
+      spins >= 0, "number of spins must be more than zero"
```
align comment with requirement. equal or more than?

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/AeronSource.scala:
```diff
@@ -92,7 +93,7 @@ private[remote] class AeronSource(
       private val sub = aeron.addSubscription(channel, streamId)
       // spin between 100 to 10000 depending on idleCpuLevel
-      private val spinning = 1100 * taskRunner.idleCpuLevel - 1000
+      private val spinning = if (inboundSourceSpinning < 0) 1100 * taskRunner.idleCpuLevel - 1000 else inboundSourceSpinning
```
We should derive it from the cpu-level and the inbound-lanes setting. We can calculate the value in ArteryTransport and pass in the parameter to the stage.

My suggestion is that we change it to:
inbound-lanes > 1 => spinning = 0

inbound-lanes == 1 =>
  cpu-level < 5 => spinning = 0
  else => something like `spinning = 50 * idleCpuLevel - 240`


The reason I want to keep the spinning is that I see a drop in the throughput when spinning=0 with inbound-lanes=1

jrudolph replied:
Ok, let's do this.

jrudolph commented on akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/FileReporter.scala:
```diff
@@ -0,0 +1,70 @@
+      }
+      val testResultFileName: String = {
+        val format = new SimpleDateFormat("yyyy-MM-dd-HH-mm-ss")
+        s"${format.format(new Date())}-Artery-$testName-$gitCommit-results.txt"
```
Hehe, yes, good point, line break is indeed weird ;)

jrudolph commented on akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/MaxThroughputSpec.scala:
```diff
@@ -174,7 +197,8 @@ object MaxThroughputSpec extends MultiNodeConfig {
+        reporter.reportResults(
           s"=== MaxThroughput ${self.path.name}: " +
             f"throughput ${throughput * testSettings.senderReceiverPairs}%,.0f msg/s, " +
             f"${throughput * payloadSize * testSettings.senderReceiverPairs}%,.0f bytes/s (payload), " +
```
This turns out quite hard to fix. The problem is that now that I changed to n-to-n communication, all the phases of the test, warmup, running and test reporting need to be synchronized over all the actors which is tricky to get right...

patriknw replied:
ok, that would require one counter per sender in the Receiver? might add to much overhead.

just report it as -1 then, if targets more than one

jrudolph replied:
The previous inaccuracy was due to warmup messages spilling into the totalReceived number. I fixed it now by wrapping warmup messages so they can be ignored for counting. (It might have a little impact on warmup as the code-paths won't be exactly the same).

jrudolph commented on akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/FileReporter.scala:
```diff
@@ -0,0 +1,70 @@
+    new FileReporter {
+      val gitCommit = {
+        import sys.process._
+        "git describe".!!
```
Good point, will fix.

jrudolph committed:
=pro fix typo

jrudolph committed:
=rem log results of MaxThroughputSpec and LatencySpec to result file

jrudolph committed:
=rem #21365 enable multiple lanes in MaxThroughputSpec

This needed the other change for each sender to send to all of the target
actors. Otherwise, large batches of messages to the same target actor would
limit the potential of actually doing work in parallel with multiple lanes due
to head-of-line blocking.

jrudolph committed:
=rem #21365 less aggressive busy spinning in AeronSource

Benchmarks revealed that busy spinning directly in the graph stage can
lead to an excessive increase in latency when multiple inbound lanes are
active (i.e. the inbound flow has an asynchronous boundary driving the
multiple lanes).

The new strategy is therefore:

For inbound-lanes > 1 or idle-cpu-level < 5: no spinning in the graph stage
For inbound-lanes = 1 and idle-cpu-level >= 6: 50 * settings.Advanced.IdleCpuLevel - 240

which means in general much less or no spinning at all.

Fixes #21365.

jrudolph commented:
Fixed all the mentioned issues and implemented the new strategy as suggested by @patriknw:

    For inbound-lanes > 1 or idle-cpu-level < 5   : no spinning in the graph stage
    For inbound-lanes = 1 and idle-cpu-level >= 6 : 50 * idle-cpu-level - 240

ktoso reviewed (approved):
LGTM :)

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/AeronSource.scala:
```diff
@@ -71,14 +71,17 @@ private[remote] object AeronSource {
   pool:           EnvelopeBufferPool,
-  flightRecorder: EventSink)
+  flightRecorder: EventSink,
+  spinning:       Int                = 0)
```
for internal api I think it's better without default value to not accidentally forget about it

jrudolph replied:
Ok, I will have to touch a few tests to do that but I guess that's fine.

patriknw reviewed (approved):
LGTM

patriknw commented:
I have started the https://jenkins.akka.io:8498/job/akka-artery-repeat/ job for this branch. Let's run it a few times to get a preview of how it behaves on jenkins

jrudolph commented:
@patriknw, thanks for starting the tests.

patriknw commented:
Remember to backport

patriknw commented:
Refs #21365
